### PR TITLE
fix: alert rule and tests association

### PR DESCRIPTION
**Pull request by brumarqu-te**
fixes https://github.com/thousandeyes/terraform-provider-thousandeyes/issues/158
This issue is happening because we currently allow the alert rule <-> test association to be established from both perspectives. From the alert rule update perspective, it will clear any test association if the list of testIds is not sent in the request.
This PR changes the alert rule `test_ids` schema attribute to be [computed](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#computed) so that it is not considered in the TF plan but it's still retrieved from the alert rule get by id operation and passed along the alert rule update request.
Also, since the alert rule returns a field `tests` when getting by id and receives a field `testIds` when updating the alert rule, had to implement a map from the GenericTest to get only the testId.

One drawback of this approach (not a huge one IMO) is that the state may not always be accurate. This is because a test <-> alert rules association can be changed from the test perspective and not get reflected on the alert rule perspective. Anyways, it will eventually update the state when the next alert rule update is performed.

### CODE REVIEW

**Comment by brumarqu-te** [comment]
/review @joaomper-TE @phpinhei-te @shahid-te @jabrowne-te @vascmart-te @mpragosa-te @sfreitas-te @pedro-te

File: thousandeyes/resource_alert_rule.go [hunk 1]
```diff
@@ -26,14 +26,20 @@ func resourceAlertRule() *schema.Resource {
...
 func resourceAlertRuleRead(d *schema.ResourceData, m interface{}) error {
 	return GetResource(d, m, func(client *thousandeyes.Client, id int64) (interface{}, error) {
-		return client.GetAlertRule(id)
+		var alerRule, err = client.GetAlertRule(id)
+		if err != nil {
+			return nil, err
+		}
+		alerRule.TestIds = testIds(*alerRule.Tests)
+		alerRule.Tests = nil
+		return alerRule, nil
```
Comment (jabrowne-te) [inline thousandeyes/resource_alert_rule.go:35]:
```suggestion
		var alertRule, err = client.GetAlertRule(id)
		if err != nil {
			return nil, err
		}
		alertRule.TestIds = testIds(*alerRule.Tests)
		alertRule.Tests = nil
		return alertRule, nil
```

**Review by pedro-te — APPROVED** [review]

File: thousandeyes/acceptance_resources/alert_rule/create_test_with_alert_rule.tf [hunk 2]
```diff
@@ -0,0 +1,34 @@
+data "thousandeyes_agent" "amsterdam" {
+  agent_name = "Amsterdam, Netherlands"
```
Comment (joaomper-TE) [inline thousandeyes/acceptance_resources/alert_rule/create_test_with_alert_rule.tf:2]: nice, more ATs for this repo! 💪

File: thousandeyes/resource_alert_rule_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+						}},
+				}
+				return testAccCheckResourceDestroy(resourceList, state)
+			},
+			checkCreateFunc: []resource.TestCheckFunc{
+				resource.TestCheckResourceAttr(alertRuleResourceName, "rounds_violating_required", "3"),
+			},
+			checkUpdateFunc: []resource.TestCheckFunc{
+				resource.TestCheckResourceAttr(alertRuleResourceName, "rounds_violating_required", "4"),
+				resource.TestCheckResourceAttr(alertRuleResourceName, "test_ids.#", "1"),
```
Comment (joaomper-TE) [inline thousandeyes/resource_alert_rule_test.go:47]: We may want to start adding some comments to these Acceptance tests - the ones that I created also do not have them and they should!
It's not immediate to understand that here we are making sure that if an `alertRule` is updated, them the `tests` associated with it are not cleared. Can you add a comment?

**Review by joaomper-TE — COMMENTED** [review]
TIL: didn't knew about this `computed` terraform attribute! This can come in handy in some edge case scenarios, such as this one

File: thousandeyes/resource_alert_rule_test.go (hunk 3 above)
Comment (brumarqu-te) [reply thousandeyes/resource_alert_rule_test.go:52]: tried to reflect that in the test name but i guess i failed miserably 😅 
i've added the comments, take another look when possible

**Review by joaomper-TE — APPROVED** [review]

**Review by jabrowne-te — APPROVED** [review]
